Sources/PackageModel/DependencyMapper.swift

```diff
@@ -29,143 +28,220 @@ public struct DefaultDependencyMapper: DependencyMapper {
             }
+            self.init(
+                parentPackagePath: parentPackagePath,
+                kind: .sourceControl(
+                    name: settings.nameForTargetDependencyResolutionOnly,
+                    location: locationString,
+                    requirement: settings.requirement
+                ),
+                productFilter: settings.productFilter
+            )
+        case .registry(let settings):
+            self.init(
+                parentPackagePath: parentPackagePath,
+                kind: .registry(
+                    id: settings.identity.description,
+                    requirement: settings.requirement
+                ),
+                productFilter: settings.productFilter
+            )
         }
+    }
+}
 
-        return nil
+extension MappablePackageDependency {
+    fileprivate var locationString: String {
+        switch self.kind {
+        case .fileSystem(_, let path):
+            return path
+        case .sourceControl(_, let location, _):
+            return location
+        case .registry(let id, _):
+            return id
+        }
     }
 
-    public static func sanitizeDependencyLocation(fileSystem: FileSystem, packageKind: PackageReference.Kind, dependencyLocation: String) throws -> String {
-        if dependencyLocation.hasPrefix("~/") {
-            // If the dependency URL starts with '~/', try to expand it.
-            return try AbsolutePath(validating: String(dependencyLocation.dropFirst(2)), relativeTo: fileSystem.homeDirectory).pathString
-        } else if dependencyLocation.hasPrefix(filePrefix) {
-            // FIXME: SwiftPM can't handle file locations with file:// scheme so we need to
-            // strip that. We need to design a Location data structure for SwiftPM.
-            let location = String(dependencyLocation.dropFirst(filePrefix.count))
-            let hostnameComponent = location.prefix(while: { $0 != "/" })
-            guard hostnameComponent.isEmpty else {
-              if hostnameComponent == ".." {
-                  throw DependencyMappingError.invalidFileURL("file:// URLs cannot be relative, did you mean to use '.package(path:)'?")
-              }
-              throw DependencyMappingError.invalidFileURL("file:// URLs with hostnames are not supported, are you missing a '/'?")
+    fileprivate var nameForTargetDependencyResolutionOnly: String? {
+        switch self.kind {
+        case .fileSystem(let name, _):
+            return name
+        case .sourceControl(let name, _, _):
+            return name
+        case .registry:
+            return .none
+        }
+    }
+
+    fileprivate var sourceControlRequirement: PackageDependency.SourceControl.Requirement {
+        get throws {
+            switch self.kind {
+            case .fileSystem:
+                throw DependencyMappingError.invalidMapping("invalid mapping of file system to source control, requirement information mismatch")
```

Comment:
Can we make this error message actionable for the user? Maybe provide expected/actual values in the error message to clarify what exactly is not matching?

Reply:
what this means is that you cannot map file system dependency to source control one as you are missing the version requirement information. do you have a suggestion as to how better say that? maybe "version requirement information not available"?

Reply:
I hope the error message could become more verbose. "Version requirement information not available" sounds better, but maybe it can provide more details where exactly it's unavailable, where it's expected to be, and how it can be provided?

Reply:
this is about mirroring depednecies. this error will be presented if someone tries to mirror file system dependency `dependency(path:)` to a source control one since there isn't version requirement on file system dependency ones. there isn't a way for the user to fix this since its invalid by definition, they need to remove it

Reply:
> perhaps something like:`mapping a file system dependency (<path>) to a source control one (<url>) is invalid`

This SGTM 👍